### PR TITLE
[Grounding DINO] Add support for cross-attention in GroundingDinoMultiHeadAttention

### LINKED ISSUE #30176: Add Cross Attention Masking for Grounding DINO
### Feature request

In the recent `GroundingDino` PR (https://github.com/huggingface/transformers/pull/26087) one thing that was missing was the cross-attention masking when running Multiheaded attention between text and image features. The assumption was that for inference the text used would probably be a fixed set of labels, thus not needing the cross-attention masking. However, for a more robust and training-friendly implementation, we should support text cross-attention mask.

### Motivation

This was an actual request in the PR of the model from people who are using the `transformers` implementation on their research see https://github.com/huggingface/transformers/pull/26087#issuecomment-2041425195

### Your contribution

A PR should be open soon 😄

### PR DESCRIPTION
**Pull request by EduardoPach**
# What does this PR do?

This PR fixes https://github.com/huggingface/transformers/issues/30176 by adding support for cross-attention masking and related integration test.

Now one can do batched inference with different text lengths such as:

```python
import torch
from transformers import AutoModelForZeroShotObjectDetection, AutoProcessor

model_id = "IDEA-Research/grounding-dino-tiny"

model = AutoModelForZeroShotObjectDetection.from_pretrained(model_id)
processor = AutoProcessor.from_pretrained(model_id)

images = [...]
text = ["a cat. a remote control.", "a person. a cat. a remote control. a couch."]

inputs = processor(images=image, text=text, padding="longest", return_tensors="pt")

with torch.no_grad():
    outputs = model(**inputs)
```

### CODE REVIEW

**Comment by amyeroberts** [comment]
Thanks for adding @EduardoPach! Let us know when it's ready for review!

**Comment by EduardoPach** [comment]
> Thanks for adding @EduardoPach! Let us know when it's ready for review!

It's already ready for review, it was easier than I was expecting

File: src/transformers/models/grounding_dino/modeling_grounding_dino.py [hunk 1]
```diff
@@ -1425,12 +1427,20 @@ def forward(
 
         # Cross-Attention Text
         queries = self.with_pos_embed(hidden_states, position_embeddings)
+        text_encoder_attention_mask = text_encoder_attention_mask[:, None, None, :]
+        text_encoder_attention_mask = text_encoder_attention_mask.repeat(
+            1, self.num_attention_heads, self.num_queries, 1
+        )
+
+        dtype = torch.float16
```
Comment (amyeroberts) [inline src/transformers/models/grounding_dino/modeling_grounding_dino.py:1435]: We shouldn't explicitly set the dtype here like this as it will break with other assumptions in the library about weight loading e.g. I should be able to do `model = AutoModel(checkpoint, torch_dtype=torch.float32)` without issue. I'd expect the dtype to be taken from e.g. the text_encoder_hidden_states or the dtype of the module - self.dtype

File: src/transformers/models/grounding_dino/modeling_grounding_dino.py [hunk 2]
```diff
@@ -1425,12 +1427,20 @@ def forward(
 
         # Cross-Attention Text
         queries = self.with_pos_embed(hidden_states, position_embeddings)
+        text_encoder_attention_mask = text_encoder_attention_mask[:, None, None, :]
```
Comment (amyeroberts) [inline src/transformers/models/grounding_dino/modeling_grounding_dino.py:1430]: This assumes it's not None

**Review by amyeroberts — COMMENTED** [review]
Thanks for adding this and a corresponding test! ❤️ 

Main comments are for robustness when using the model in transformers

File: src/transformers/models/grounding_dino/modeling_grounding_dino.py (hunk 1 above)
Comment (EduardoPach) [reply src/transformers/models/grounding_dino/modeling_grounding_dino.py]: Ops, this was a copy-pasta! Thanks for spotting this. Fixed here https://github.com/huggingface/transformers/pull/30364/commits/539763d21cb6469267ef7b54bdbf085d5c90c7b1 for the cross-attention and for the text enhancer that had the same logic

File: src/transformers/models/grounding_dino/modeling_grounding_dino.py (hunk 2 above)
Comment (EduardoPach) [reply src/transformers/models/grounding_dino/modeling_grounding_dino.py]: https://github.com/huggingface/transformers/pull/30364/commits/ca3933d97808ebff556309486b0514d8d9db3d30

File: src/transformers/models/grounding_dino/modeling_grounding_dino.py [hunk 3]
```diff
@@ -1425,12 +1427,21 @@ def forward(
...
         queries = self.with_pos_embed(hidden_states, position_embeddings)
+        if text_encoder_attention_mask is not None:
+            dtype = text_encoder_hidden_states.dtype
+
+            text_encoder_attention_mask = text_encoder_attention_mask[:, None, None, :]
+            text_encoder_attention_mask = text_encoder_attention_mask.repeat(
+                1, self.num_attention_heads, self.num_queries, 1
+            )
+            text_encoder_attention_mask = text_encoder_attention_mask.to(dtype=dtype)
+            text_encoder_attention_mask = text_encoder_attention_mask * torch.finfo(dtype).min
```
Comment (amyeroberts) [inline src/transformers/models/grounding_dino/modeling_grounding_dino.py:1438]: Actually - just realised this is in the `DecoderLayer`. This should be done in the decoder, to avoid recreating the repeated attention mask for every decoder layer.

Comment (EduardoPach) [reply src/transformers/models/grounding_dino/modeling_grounding_dino.py]: https://github.com/huggingface/transformers/pull/30364/commits/0a520b1f5ba9b57938ffc9d2cf6d38575bbb0b6b

**Review by amyeroberts — APPROVED** [review]
Thanks for adding!
